Feature/create desk reservation

MonikaLukowska opened the pull request:
This PR adds create at desk reservation. Implementation includes:

- routes
- controller
- usecase
- test

I am aware that this use case is very similar to create use case but I decided to separate them because of 2 main reasons:

- I can avoid some conditional logic in few places in application
- In my opinion code is cleaner and easier to read

MonikaLukowska committed:
chore:update git ignore

MonikaLukowska committed:
Merge branch 'main' of github.com:MonikaLukowska/monte-cinema

MonikaLukowska committed:
feat:add routes for create at desk reservation

MonikaLukowska committed:
feat:add create at desk use case and controller action

MonikaLukowska committed:
feat:add create at desk policy

MonikaLukowska committed:
feat:update views

MonikaLukowska committed:
test: add create at desk spec

MonikaLukowska committed:
refactor: clean up reservation controller

rrothenberger commented on app/concepts/reservations/use_cases/create.rb:
```diff
@@ -49,7 +49,7 @@ def deadline
 
       def not_too_late?
-        return true if user.manager? || (user.client? && seance.start_time > deadline)
+        return true if seance.start_time > deadline
```
I have a small suggestion that will probably help with code readability:

```ruby
def is_after_time_limit?
  seance.start_time - Reservation::CONFIRMATION_DEADLINE < DateTime.current
end

def not_too_late?
  return true if is_after_time_limit?

  errors << 'This seance starts in 30 minutes or less, make reservation at ticket desk'
  false
end
```

But even if you don't follow it, please change `DateTime.now` to `DateTime.current`.
`DateTime.now` will return time with timezone of machine it is running.
`DateTime.current` will return time with timezone set in rails configuration.

Here simple example of the difference:
![image](https://user-images.githubusercontent.com/36715675/194608323-d25f5117-a999-4ca6-849d-bc5067508b2e.png)
As you can see, the first result is with the timezone specified as +0200 - which translates to 2 hours ahead UTC. Which makes sense - Europe/Warsaw is currently following CEST, which is 2 hours ahead of UTC, and that's timezone I have on my PC. But if I would deploy that app in, let's say Romania, it could have timezone there set to +0300 - which DateTime.now would return different time.
And in seconds result we have UTC +0000, so just UTC - which is set in my `application.rb`:
![image](https://user-images.githubusercontent.com/36715675/194611300-028e56c8-c41d-43b8-bcc9-038fa9683c03.png)

Why is this important?
One reason I want to talk about is predictability - `DateTime.current` will always return timezone that your application is configured to use - so you will have configured timezone always returned, which is not the case with `DateTime.now` - here it depends on the machine that is running specific instance - and so we enter the world of problems. You might think that as long as timezone of server matches the timezone of users everything is golden - but that's not always possible. None of platforms for deployments we showed you allows us to deploy in Poland - fortunately, we might deploy in Germany, which has time zone, right? Well yeah - but as we know are talking about abandoning changing between winter and summer time in Europe - it could break - because maybe they will be first to drop that change, and we fill follow year or two later. Which would lead to mismatches in timezones and we lose that predictability. And this assumes that the timezone used on the runner in heroku/render/railway actually is set to local time for datacenter - which might not be the case, often on runners UTC is set as kind of "widely accepted default timezone". And even if they use local time, what if it will be not deployed in Germany, but somewhere else where daylight saving times are not a thing, or change on different dates.

Another thing is consistency - for consistency it would be great to as quickly as possible to convert on BE everything to common preferable not changing(so no daylight saving times) timezone. Again, there is industry consensus to use for that UTC. Having consistency means that you always know what you are working with and makes it easier to create and maintain tests. And why use `DateTime.current` over `DateTime.now` everywhere - that would be consistent too(assuming that we have perfect control over the deployment environment)? Because of everything in the ecosystem that already uses the timezone of the application - eg. all timestamps in ActiveRecord - you can quickly check that `created_at` should return time in the timezone specified in the application.rb 😉

MonikaLukowska replied:
Dated and timezones - my this week leitmotiv 😄 Thanks for comprehensive explanation, it was very helpful. I will apply these suggestions.

rrothenberger commented on app/concepts/reservations/use_cases/create_at_desk.rb:
```diff
@@ -0,0 +1,49 @@
+      def call
+        return unless seats_selected?
+
+        ActiveRecord::Base.transaction do
+          reservation.tap do |reservation|
+            seats.each do |seat|
+              Ticket.create!(reservation_id: reservation.id, seat: seat)
+            end
+          end
+        end
+        reservation
```
```suggestion
        Reservation.create!(
          user: user,
          seance_id: seance.id,
          status: Reservation::CONFIRMED,
          tickets: seats.map { |seat| Ticket.new(seat: seat) }
        )
```

This has two benefits - first, you don't need manually call transaction, as ActiveRecord will handle that now for you. And it should return in two queries - one for creation of reservation and one for creation of all tickets at the same time. Where yours approach creates a query per ticket.

MonikaLukowska replied:
Nice, I did not know I could do something like this. What makes it act like a transaction?

rrothenberger replied:
Because that's a single function call - `Ticket.new` does not yet create ticket in database, it just initialize the object. Then when `Reservation.create!` is called with it will then create both reservation and tickets -  and ActiveRecord will put all of that in transaction by default 😄

MonikaLukowska replied:
clever, I like it! :)

rrothenberger commented:
btw. nice small PR, good job 🎉

MonikaLukowska committed:
refactor:apply CR remarks

change way of creating the reservation, add after_time_limit method, use DateTime.current instead od DateTime.now

MonikaLukowska committed:
refactor: change DateTime.now to DateTime.current

MonikaLukowska committed:
fix:use correct use case in controller actions
